### PR TITLE
Fix stubgen and package stubs in wheel/sdists

### LINKED ISSUE #2798: Decide best way to ship `pyi` files with sdist/wheel
## 🧰 Task
Once #2675 is merged, developers will be able to get type stubs for a couple autogenerated modules by running `make typestubs` in their dev environment.  We still need to figure out the best way to ship those `.pyi` with the sdist...

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
follows on #2675 and closes #2798 
this includes our autogenerated types stubs in the sdist and wheel

i tried to get it to work with `pip install -e .`, including everything in [this post](https://stackoverflow.com/a/45021666), but failed so far.  So devs may still need to run `make typestubs` locally.

also, for this to work with `python -m build` (build [package](https://github.com/pypa/build)) in our `make_release.yml`, we'd need to duplicate our dependencies in `pyproject.toml`, so i reverted back to using `python setup.py sdist bdist_wheel`.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/stubgen.py
Comment (tlambert03) [inline napari/utils/stubgen.py]: this is really quite ugly... but this seems to be the cost of using things like `__future__.annotations` and `TYPE_CHECKING` imports. we need to re-import modules with `typing.TYPE_CHECKING == True`, which only sometimes works.  but for now, it seems to be "good enough"

**Comment by jni** [comment]
Copying an explanation @tlambert03 offered me on Zulip for this PR, because I think it's important background here:

--------

ok. so, running make typestubs was failing again, and it fails because of this forward ref right here: https://github.com/napari/napari/blob/82ca467b515023a9ad6105cfbace9282f67be69c/napari/components/viewer_model.py#L796

```python
    def open(
        self,
        path: PathOrPaths,
        ...
```

when we try to build `view_layers.py`, the `napari.open` function pulls from `napari.viewer`, which pulls from `napari.components.viewer_model`

and that file has a forwardRef type hint that is hidden behind

```python
if TYPE_CHECKING:
    from ..types import FullLayerData, LayerData, PathOrPaths
```

if you try to import directly, you get circular import...

so you need to import, turn on `TYPE_CHECKING`, and reimport
but, you can't just say "do it for this current module" ... because... modules all the way down

(@jni: but why do you need to reload, rather than just import?)

because it's already been imported with `TYPE_CHECKING = False`

and `PathOrPaths` isn't in `module.__globals__`

which is what get_type_hints needs to resolve the forward ref :)

**Comment by jni** [comment]
Hence hacking importing/reloading.

I'm not in love with the current implementation. @tlambert03 offered the option of checking in type stubs and making sure that they are up to date using CI, which he says is what most projects are doing. Though I'm still generally reluctant to check in generated files, it seems preferable to this hack, *I think*. Curious to hear others' opinions here.

**Comment by tlambert03** [comment]
(And the last option was to just punt on the hack, and change the one problematic type hint... and let it break again in the future)

**Comment by tlambert03** [comment]
ok @jni, have another look.  I did two things
1) just manually declare PathOrPaths again in `viewer_model.py` to avoid importing from types
2) manually build typestubs during `make_release.yml` (via the `make dist` command).

**Review by sofroniewn — APPROVED** [review]
This looks fine to me, but given comments above I'll wait for @jni to weigh in before merge

**Review by jni — APPROVED** [review]
Oh man this is so much better now @tlambert03! 😅 

Can you fix the merge conflicts in the Makefile? I had a look and it wasn't suuuuper obvious what needed to be done so I left it. 😬 But feel free to self-merge after. 🎉

**Comment by tlambert03** [comment]
> Oh man this is so much better now @tlambert03! 😅

great.  I agree it's better... but it will also fail if unresolvable forward refs make their way back into these modules.  I removed the try/catch from the stubgen `__main__` section so it will at least fail loudly!

**Pull request merged**
